packages/api/src/config.js

```diff
@@ -0,0 +1,321 @@
+ * used if a variable is missing.
+ *
+ * @returns {ServiceConfiguration}
+ */
+export function loadServiceConfig() {
+  const vars = loadConfigVariables()
+  mergeDefaultValuesIfAllowed(vars)
+
+  return serviceConfigFromVariables(vars)
+}
+
+/**
+ * Parse a {@link ServiceConfiguration} out of the given `configVars` map.
+ * @param {Record<string, string>} configVars map of variable names to values.
+ *
+ * Exported for testing. See {@link getServiceConfig} for main public accessor.
+ *
+ * @returns {ServiceConfiguration}
+ */
+export function serviceConfigFromVariables(configVars) {
+  let clusterUrl = configVars.CLUSTER_API_URL
+  if (configVars.CLUSTER_SERVICE) {
+    const serviceUrl = CLUSTER_SERVICE_URLS[configVars.CLUSTER_SERVICE]
+    if (!serviceUrl) {
+      throw new Error(`unknown cluster service: ${configVars.CLUSTER_SERVICE}`)
+    }
+    clusterUrl = serviceUrl
+  }
+
+  return {
+    isDebugBuild: boolValue(configVars.DEBUG),
+    runtimeEnvironment: runtimeEnvFromString(configVars.ENV),
+    version: {
+      semver: configVars.VERSION,
+      branch: configVars.BRANCH,
+      commitHash: configVars.COMMITHASH,
+    },
+
+    maintenanceMode: maintenanceModeFromString(configVars.MAINTENANCE_MODE),
+
+    secrets: {
+      salt: configVars.SALT,
+      metaplexAuthToken: configVars.METAPLEX_AUTH_TOKEN,
+      ucanPrivateKey: configVars.PRIVATE_KEY,
+    },
+
+    external: {
+      cluster: {
+        url: clusterUrl,
+        basicAuthToken: configVars.CLUSTER_BASIC_AUTH_TOKEN,
+      },
+
+      database: {
+        url: configVars.DATABASE_URL,
+        authToken: configVars.DATABASE_TOKEN,
+      },
+
+      s3: {
+        endpoint: configVars.S3_ENDPOINT,
+        region: configVars.S3_REGION,
+        accessKeyId: configVars.S3_ACCESS_KEY_ID,
+        secretAccessKey: configVars.S3_SECRET_ACCESS_KEY,
+        bucketName: configVars.S3_BUCKET_NAME,
+      },
+
+      magicLink: {
+        secret: configVars.MAGIC_SECRET_KEY,
+      },
+
+      logtail: {
+        authToken: configVars.LOGTAIL_TOKEN,
+      },
+
+      sentry: {
+        dsn: configVars.SENTRY_DSN,
+      },
+
+      mailchimp: {
+        apiKey: configVars.MAILCHIMP_API_KEY,
+      },
+    },
+  }
+}
+
+/**
+ * Loads configuration variables from the global environment and returns a JS object
+ * keyed by variable names.
+ *
+ * Exported for testing. See {@link getServiceConfig} for main config accessor.
+ *
+ * @returns { Record<string, string>} an object with `vars` containing all found variables and their values, and a `notFound` array of variables that were not defined
+ */
+export function loadConfigVariables() {
+  /** @type Record<string, string> */
+  const vars = {}
+
+  /** @type Record<string, unknown> */
+  const globals = globalThis
+
+  for (const name of Object.keys(DEFAULT_CONFIG_VALUES)) {
+    const val = globals[name]
+    if (typeof val === 'string') {
+      vars[name] = val
+    }
+  }
```

Comment:
I think we should do more in this for loop and remove some of the other functions.

Merging should happen right here for `test` and `dev`, fail for the other env. Basically the some logic overall just in one place, makes it easier to flow.

Reply:
Also we avoid iterating multiple times over the config keys and we can make it all static and return one single object from this file.